File: juriscore-billing/src/test/java/com/juriscore/billing/domain/MoneyTest.java
package com.juriscore.billing.domain;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import java.math.BigDecimal;
import java.math.RoundingMode;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * The arithmetic an invoice is built on.
 *
 * <p>Asserted directly and in detail, because this is the one place in the platform where
 * being subtly wrong costs a firm money rather than an error message.
 */
class MoneyTest {

    @Test
    @DisplayName("every currency amount lands on two decimals")
    void amountsCarryTwoDecimals() {
        assertThat(Money.amount(new BigDecimal("5"))).isEqualTo(new BigDecimal("5.00"));
        assertThat(Money.amount(new BigDecimal("5.1"))).isEqualTo(new BigDecimal("5.10"));
        assertThat(Money.amount(new BigDecimal("5.005"))).isEqualTo(new BigDecimal("5.01"));
        assertThat(Money.amount(null)).isEqualTo(new BigDecimal("0.00"));
    }

    @Test
    @DisplayName("HALF_UP, which is what a person does by hand")
    void roundsHalvesAway() {
        assertThat(Money.ROUNDING).isEqualTo(RoundingMode.HALF_UP);
        // HALF_EVEN would give 2.12 here, and a firm explaining that to a client is a
        // conversation that should not have to happen.
        assertThat(Money.amount(new BigDecimal("2.125"))).isEqualTo(new BigDecimal("2.13"));
        assertThat(Money.amount(new BigDecimal("2.135"))).isEqualTo(new BigDecimal("2.14"));
    }

    @ParameterizedTest(name = "{0} × {1} = {2}")
    @CsvSource({
            "1.000,   0.01,        0.01",
            "3.000,   0.01,        0.03",
            "0.333,   100.00,     33.30",
            "2.500,   4000.00, 10000.00",
            "1.005,   100.00,    100.50",
            "0.001,     0.01,      0.00",
            "7.000, 999999.99, 6999999.93"})
    @DisplayName("a line amount is the rounded product of quantity and unit price")
    void lineAmounts(String quantity, String unitPrice, String expected) {
        assertThat(Money.lineAmount(new BigDecimal(quantity), new BigDecimal(unitPrice)))
                .isEqualByComparingTo(new BigDecimal(expected));
    }

    @Test
    @DisplayName("a hundredth of a rupee survives; nothing silently rounds to zero above it")
    void thePennyCase() {
        assertThat(Money.lineAmount(new BigDecimal("1.000"), new BigDecimal("0.01")))
                .isEqualByComparingTo(new BigDecimal("0.01"));
        // Below the smallest unit there is nothing to keep, and saying so is the honest
        // behaviour: 0.001 × 0.01 really is less than half a paisa.
        assertThat(Money.lineAmount(new BigDecimal("0.001"), new BigDecimal("0.01")))
                .isEqualByComparingTo(BigDecimal.ZERO);
    }

    @ParameterizedTest(name = "{1}% of {0} = {2}")
    @CsvSource({
            "10000.00, 18.000, 1800.00",
            "  100.00, 18.000,   18.00",
            "    0.05, 18.000,    0.01",
            "    0.01, 18.000,    0.00",
            "  999.99,  5.500,   55.00",
            " 1234.56, 12.345,  152.41",
            "  100.00,  0.000,    0.00"})
    @DisplayName("tax is a percentage of the already-rounded line amount")
    void taxRounding(String amount, String rate, String expected) {
        assertThat(Money.taxOn(new BigDecimal(amount), new BigDecimal(rate)))
                .isEqualByComparingTo(new BigDecimal(expected));
    }

    @Test
    @DisplayName("a null or zero rate costs nothing to compute")
    void noTaxIsZero() {
        assertThat(Money.taxOn(new BigDecimal("100.00"), null)).isEqualByComparingTo(BigDecimal.ZERO);
        assertThat(Money.taxOn(new BigDecimal("100.00"), BigDecimal.ZERO))
                .isEqualByComparingTo(BigDecimal.ZERO);
    }

    @Test
    @DisplayName("summing rounded amounts is exact, which is why the invoice total is an identity")
    void sumsOfRoundedAmountsDoNotDrift() {
        // The floating-point disaster case: 0.1 + 0.2 != 0.3 in a double.
        BigDecimal total = Money.ZERO;
        for (int i = 0; i < 3; i++) {
            total = total.add(Money.amount(new BigDecimal("0.10")));
        }
        assertThat(total).isEqualByComparingTo(new BigDecimal("0.30"));

        // A hundred lines of a third of a rupee, each rounded, sums to exactly what the
        // printed invoice says — not to something a hundredth away from it.
        BigDecimal hundredThirds = Money.ZERO;
        for (int i = 0; i < 100; i++) {
            hundredThirds = hundredThirds.add(Money.lineAmount(
                    new BigDecimal("1.000"), new BigDecimal("0.33")));
        }
        assertThat(hundredThirds).isEqualByComparingTo(new BigDecimal("33.00"));
    }

    @Test
    @DisplayName("large totals stay exact — no precision cliff on the way up")
    void largeTotals() {
        BigDecimal huge = Money.lineAmount(new BigDecimal("1000.000"), new BigDecimal("99999.99"));
        assertThat(huge).isEqualByComparingTo(new BigDecimal("99999990.00"));
        assertThat(Money.sum(huge, new BigDecimal("0.01")))
                .isEqualByComparingTo(new BigDecimal("99999990.01"));
    }

    @Test
    void recognisesSignAndEquality() {
        assertThat(Money.isPositive(new BigDecimal("0.01"))).isTrue();
        assertThat(Money.isPositive(BigDecimal.ZERO)).isFalse();
        assertThat(Money.isPositive(null)).isFalse();
        assertThat(Money.isNegative(new BigDecimal("-0.01"))).isTrue();
        assertThat(Money.equal(new BigDecimal("5"), new BigDecimal("5.00"))).isTrue();
        assertThat(Money.equal(new BigDecimal("5.00"), new BigDecimal("5.01"))).isFalse();
    }

    @Test
    @DisplayName("quantities carry three decimals and rates three, and both are documented")
    void scales() {
        assertThat(Money.SCALE).isEqualTo(2);
        assertThat(Money.QUANTITY_SCALE).isEqualTo(3);
        assertThat(Money.RATE_SCALE).isEqualTo(3);
        assertThat(Money.quantity(new BigDecimal("2.5"))).isEqualTo(new BigDecimal("2.500"));
        assertThat(Money.rate(new BigDecimal("18"))).isEqualTo(new BigDecimal("18.000"));
        assertThat(Money.rate(null)).isEqualByComparingTo(BigDecimal.ZERO);
    }
}

File: frontend/src/lib/money.test.ts
import { describe, expect, it } from 'vitest';
import {
  estimateTotals, multiply, parseDecimal, percentOf, roundToCurrency, toDecimalString,
} from './money';

describe('parseDecimal', () => {
  it('reads plain decimals at full precision', () => {
    expect(toDecimalString(parseDecimal('1234.56') ?? 0n)).toBe('1234.56');
    expect(toDecimalString(parseDecimal('.5') ?? 0n)).toBe('0.50');
    expect(toDecimalString(parseDecimal('7.') ?? 0n)).toBe('7.00');
    expect(toDecimalString(parseDecimal('-12.5') ?? 0n)).toBe('-12.50');
  });

  it('rejects anything that is not a plain decimal', () => {
    for (const bad of ['', ' ', '.', '-', 'abc', '1e5', '1,000', '1.2.3', '12px']) {
      expect(parseDecimal(bad), bad).toBeNull();
    }
  });
});

describe('exact arithmetic', () => {
  /**
   * The reason this module exists at all. In IEEE-754 doubles `0.1 + 0.2` is
   * 0.30000000000000004 and `1.005 * 100` is 100.49999999999999; neither is acceptable in
   * a figure a lawyer reads next to a client's name.
   */
  it('does not accumulate binary floating-point drift', () => {
    const tenth = parseDecimal('0.1') ?? 0n;
    const fifth = parseDecimal('0.2') ?? 0n;
    expect(toDecimalString(tenth + fifth)).toBe('0.30');
    expect(0.1 + 0.2).not.toBe(0.3); // the behaviour being avoided

    const rate = parseDecimal('1.005') ?? 0n;
    const hundred = parseDecimal('100') ?? 0n;
    expect(toDecimalString(multiply(rate, hundred))).toBe('100.50');
  });

  it('rounds half away from zero, matching Java HALF_UP', () => {
    expect(toDecimalString(roundToCurrency(parseDecimal('2.005') ?? 0n))).toBe('2.01');
    expect(toDecimalString(roundToCurrency(parseDecimal('2.015') ?? 0n))).toBe('2.02');
    expect(toDecimalString(roundToCurrency(parseDecimal('2.004') ?? 0n))).toBe('2.00');
    expect(toDecimalString(roundToCurrency(parseDecimal('-2.005') ?? 0n))).toBe('-2.01');
  });

  it('computes a percentage of an amount', () => {
    const amount = parseDecimal('11800') ?? 0n;
    const rate = parseDecimal('18') ?? 0n;
    expect(toDecimalString(percentOf(amount, rate))).toBe('2124.00');
  });
});

describe('estimateTotals', () => {
  it('taxes each line separately, as the backend does', () => {
    // Two lines at different rates: taxing the combined subtotal would give a different
    // answer, which is exactly the mistake this mirrors the backend to avoid.
    const totals = estimateTotals([
      { quantity: '2', unitPrice: '5000', taxRate: '18' },
      { quantity: '1', unitPrice: '2000', taxRate: '5' },
    ], '');

    expect(totals.subtotal).toBe('12000.00');
    expect(totals.taxAmount).toBe('1900.00'); // 1800 + 100
    expect(totals.total).toBe('13900.00');
    expect(totals.partial).toBe(false);
  });

  it('subtracts a flat discount from the taxed total', () => {
    const totals = estimateTotals(
      [{ quantity: '1', unitPrice: '1000', taxRate: '10' }],
      '250',
    );
    expect(totals.total).toBe('850.00');
  });

  it('treats a blank tax rate as zero', () => {
    const totals = estimateTotals([{ quantity: '3', unitPrice: '99.99', taxRate: '' }], '');
    expect(totals.subtotal).toBe('299.97');
    expect(totals.taxAmount).toBe('0.00');
    expect(totals.partial).toBe(false);
  });

  it('flags a half-typed line rather than silently counting it as zero', () => {
    const totals = estimateTotals([
      { quantity: '1', unitPrice: '500', taxRate: '' },
      { quantity: '2', unitPrice: '', taxRate: '' },
    ], '');

    expect(totals.subtotal).toBe('500.00');
    expect(totals.partial).toBe(true);
  });
});
